### PR TITLE
GNA plugin transformations documentation

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
 - add md file with transformations documentation

### Tickets:
 - n/a

### CODE REVIEW

File: src/plugins/intel_gna/docs/transformations.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# GNA transformations documentation
+
+GNAPlugin has multiple methods, one of the most important is LoadNetwork. That method prepares input graph to execute on the GNA hardware. Input graph usually consists of variety of operations, i.e. Convolution, Add, Gather, LSTMSequence. GNA hardware provides limited number of operations. First of all LoadNetwork method converts input network to equivalent network with layers each of them could be executed on the GNA hardware.
```
Comment (marcinkusm) [inline src/plugins/intel_gna/docs/transformations.md:3]: GNA Plugin provides implementation of multiple methods required by OpenVINO plugin API. Input graph usually consists of variety of operations, i.e. Convolution, Add, Gather, LSTMSequence and so on. GNA Hardware is its own limitation and not all operations can be executed on GNA Hardware.
One of the main functionalities for GNA Plugin is conversion of input network to equivalent network which could be executed on the GNA hardware. This conversion is realized by LoadNetwok method.

File: src/plugins/intel_gna/docs/transformations.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+There is a work moving entire GNA plugin from legacy CNNNetwork API to the ngraph.
+
+## LoadNetwork
+
+GNAPlugin::LoadNetwork will start with the next stages:
```
Comment (marcinkusm) [inline src/plugins/intel_gna/docs/transformations.md:8]:
```
GNAPlugin::LoadNetwork will start with the next stages:
-	Converting input graph to fully GNA-supported graph (all in ngraph)
-	Creating and connecting GNA primitives within libGNA from ngraph-based network
```
If this chapter means how LoadNetwork  should work in the future:
 I would change `GNAPlugin::LoadNetwork will start with the next stages:` to
GNAPlugin::LoadNetwork in the future should execute following stages:
-	Converting input graph to fully GNA-supported graph (all in ngraph)
-	Creating and connecting GNA primitives within libGNA from ngraph-based network

File: src/plugins/intel_gna/docs/transformations.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+GNA-specific ngraph-based transformations are placed in src/plugins/intel_gna/src/transformations. All transformations should have brief comments in it’s header. That brief should describe what pattern transformation handles and what modifications doing.
+There is also a directory src/plugins/intel_gna/src/transformations/rt_info with auxiliary runtime attributes. That attributes could be added into node rt_info map. That attributes can be read/write in transformations which is useful in some cases. For example, transformation can proceed some node in a special way if the node has special attribute.
+All new transformations should have unit tests, that are placed in src/plugins/intel_gna/tests/unit/transformations. All that unit tests are compiled in ov_gna_unit_tests binary.
+CNNNetwork transformations are so-called passes. They are placed in src/plugins/intel_gna/src/optimizer/gna_pass_manager.cpp. Passes proceed network as a
+```
+std::vector<InferenceEngine::CNNLayerPtr> * pLayers
+```
+It is preferrable not to write new CNNNetwork transformations. All that code is considered as a legacy. Existed CNNNetwork passes are ported to ngraph.
```
Comment (marcinkusm) [inline src/plugins/intel_gna/docs/transformations.md:28]:
```suggestion
It is preferrable to write new transformations as nGraph passes and avoid implementing CNNNetwork passes. All CNNNetwork related code is considered as a legacy. Existed CNNNetwork passes are ported to ngraph.
```

**Review by marcinkusm — COMMENTED** [review]
this is partial review, will continue on Monday.

File: src/plugins/intel_gna/docs/transformations.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+TransposeSinking changes Concat and Split axis while pushing Transpose nodes through them. GNA is not able to execute Concat/Split operations with all axis values. Some TransposeSinking transformations supports callbacks. That callbacks are executed inside transformations and allows to add plugin specific checks. GNA plugin prevents sinking transposes in these checks that changes Split/Concat layers from supported to unsupported by GNA plugin.
```
Comment (rjeziers) [inline src/plugins/intel_gna/docs/transformations.md:47]:
```suggestion
TransposeSinking changes Concat and Split axis while pushing Transpose nodes through them. GNA doesn't support all possible Concat and Split axis. Some TransposeSinking transformations support callbacks. These callbacks are executed inside transformations and allow to add plugin specific checks. In these checks, GNA plugin prevents sinking transposes that would make some Split/Concats unsupported.
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+One of the main functionalities for GNA Plugin is conversion of input network to equivalent network which could be executed on the GNA hardware. This conversion is realized by LoadNetwok method.
+
+## LoadNetwork
+
+GNAPlugin::LoadNetwork in the future should execute following stages:
+-	Converting input graph to fully GNA-supported graph (all in ngraph)
+-	Creating and connecting GNA primitives within libGNA from ngraph-based network
+
+These stages include:
+-	Obtaining ngraph-based network from the CNNNetwork argument (if input is not ngraph-based, proceed to CNNNetwork passes stage)
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:13]:
```suggestion
-	Obtain ngraph-based network from the CNNNetwork argument (if input is not ngraph-based, proceed to CNNNetwork passes stage)
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+1.	src/plugins/intel_gna/legacy/include/legacy/ngraph_ops
+Here there are legacy layer types. Their names ends with “IE”. These types cannot be in graph, that pass to GNA plugin. All of these types are created within GNA transformations and used in GNA graph compiler for creating libGNA primitives. There are plans to rewrite all legacy code. These legacy types should be removed after that.
+2.	src/plugins/intel_gna/src/ops
+GNA-specific operations. For example, GNAConvolution type describes convolution layers. It differs from common OpenVino Convolution type as it handles NHWC data layout instead of NCHW.
+Ngraph-based transformations
+1.	Transformations that are common for all OpenVino plugins (are placed outside GNA plugin directory). These transformations perform different optimizations. For example, ov::pass::ConvertDivide transforms Divide operation into the sequence of node with Power layer. LSTMCellDecomposition extracts LSTMCell into subgraph of mathematical operations.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:38]:
```suggestion
1.	Transformations that are common for all OpenVino plugins (are placed outside GNA plugin directory). These transformations perform different optimizations. For example, ov::pass::ConvertDivide transforms Divide operation into the sequence of nodes with Power layer. LSTMCellDecomposition extracts LSTMCell into subgraph of mathematical operations.
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+All new transformations should have unit tests, that are placed in src/plugins/intel_gna/tests/unit/transformations. All that unit tests are compiled in ov_gna_unit_tests binary.
+CNNNetwork transformations are so-called passes. They are placed in src/plugins/intel_gna/src/optimizer/gna_pass_manager.cpp. Passes proceed network as a
+```
+std::vector<InferenceEngine::CNNLayerPtr> * pLayers
+```
+It is preferrable to write new transformations as nGraph passes and avoid implementing CNNNetwork passes. All CNNNetwork related code is considered as a legacy. Existed CNNNetwork passes are ported to ngraph.
+
+## GNA ngraph-based layers
+
+OpenVino allows to work with graph nodes as ov::Node class instances. Most of them stores in src/core/include/openvino/op directory and could be used by all plugins. GNA plugin stores own (GNA-specific) layer types.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:32]:
```suggestion
OpenVino allows to work with graph nodes as ov::Node class instances. Most of them are stored in src/core/include/openvino/op directory and could be used by all plugins. GNA plugin stores own (GNA-specific) layer types.
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+-	Converting input graph to fully GNA-supported graph (all in ngraph)
+-	Creating and connecting GNA primitives within libGNA from ngraph-based network
+
+These stages include:
+-	Obtain ngraph-based network from the CNNNetwork argument (if input is not ngraph-based, proceed to CNNNetwork passes stage)
+-	Pass ngraph-based network through ngraph-based transformations.
+-	Convert ngraph-based network to CNNNetwork-based
+-	Pass network through CNNNetwork-based transformations.
+-	Creating and connecting GNA primitives withing libGNA from CNNNetwork-bases graph
+Transformations are the way of modifying input graph. Ngraph-based transformations usually are the next types:
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:18]:
```suggestion
Transformations are the way of modifying input graph. Ngraph-based transformations usually are of the following types:
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+-	inherited from ov::pass::ModelPass. They implement run_on_model method. It allows them to be a container of other transformations. For example, ngraph::pass::CommonOptimizations executes multiple transformations in it. Each of them do some basic transormations.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:19]:
```suggestion
-	inherited from ov::pass::ModelPass. They implement run_on_model method. It allows them to be a container of other transformations. For example, ngraph::pass::CommonOptimizations executes multiple transformations in it. Each of them do some basic transformations.
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+-	inherited from ov::pass::MatcherPass. Such transformations usually have a constructor. That constructor defines a pattern with the several connected together layers and a function that modifies found group of layers. The pattern can also handle additional predicates that do any checks on the traversed nodes. It is preferable to use that predicates explicitly rather than check and return from the transform function.
+GNA-specific ngraph-based transformations are placed in src/plugins/intel_gna/src/transformations. All transformations should have brief comments in it’s header. That brief should describe what pattern transformation handles and what modifications doing.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:21]:
```suggestion
GNA-specific ngraph-based transformations are placed in src/plugins/intel_gna/src/transformations. All transformations should have brief comments in their headers. That brief should describe what pattern transformation handles and what modifications do.
```

**Review by mryzhov — APPROVED** [review]

File: src/plugins/intel_gna/docs/transformations.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+-	Execute a set of transformations to convert Slice -> StridedSlice -> CropIE nodes
+When StridedSlice layer will be supported by TransposeSInking these steps could be removed from GNA plugin pipeline.
+
+## CNNNetwork based passes
+
+After running ngraph-based transformations model is converted with function convertFunctionToICNNNetwork into CNNNetwork-based function. The next step is the model transformation with the CNNNetwork-based passes.
+All the legacy CNNNetwork-based passes are stored in src/plugins/intel_gna/src/optimizer/gna_pass_manager.cpp. One of the main difference between legacy passes and ngraph transformations is that legacy passes doesn’t have pattern matching functionality. Each of the passes iterating through the graph nodes (previously sorting toplogical) searching for sought sequence of layers and modify them.
+It should be mentioned that ngraph API stores constant data as input nodes with type Constant, but CNNNetwork API stores data as a BLOB in layer info.
+
+## Debugging
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/plugins/intel_gna/docs/transformations.md:67]: yes, this should be a part of speech sample documentation

File: src/plugins/intel_gna/docs/transformations.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
+# GNA transformations documentation
+
+GNA Plugin provides implementation of multiple methods required by OpenVINO plugin API. Input graph usually consists of variety of operations, i.e. Convolution, Add, Gather, LSTMSequence and so on. GNA Hardware is its own limitation and not all operations can be executed on GNA Hardware.
+One of the main functionalities for GNA Plugin is conversion of input network to equivalent network which could be executed on the GNA hardware. This conversion is realized by LoadNetwok method.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:4]:
```suggestion
One of the main functionalities for GNA Plugin is conversion of source network to equivalent network which could be executed on the GNA hardware. This conversion is done in LoadNetwok method.
```

File: src/plugins/intel_gna/docs/transformations.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
+# GNA transformations documentation
+
+GNA Plugin provides implementation of multiple methods required by OpenVINO plugin API. Input graph usually consists of variety of operations, i.e. Convolution, Add, Gather, LSTMSequence and so on. GNA Hardware is its own limitation and not all operations can be executed on GNA Hardware.
```
Comment (mryzhov) [inline src/plugins/intel_gna/docs/transformations.md:3]:
```suggestion
GNA Plugin provides implementation of multiple methods required by OpenVINO plugin API. Original model usually consists of variety of operations, i.e. Convolution, Add, Gather, LSTMSequence and so on. GNA Hardware is its own limitation and not all operations can be executed on GNA Hardware.
```

**Pull request merged**